extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -212,44 +214,119 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
                 continue;
             }
             BindableServiceBuildItem item = new BindableServiceBuildItem(service.name());
-            Set<MethodInfo> blockingMethods = gatherBlockingMethods(service);
-            for (MethodInfo method : blockingMethods) {
-                item.registerBlockingMethod(method.name());
+            Set<String> blockingMethods = gatherBlockingMethods(service, index);
+            for (String method : blockingMethods) {
+                item.registerBlockingMethod(method);
             }
             bindables.produce(item);
         }
     }
 
-    private Set<MethodInfo> gatherBlockingMethods(ClassInfo service) {
-        Set<MethodInfo> result = new HashSet<>();
-        boolean blockingClass = service.classAnnotation(GrpcDotNames.BLOCKING) != null;
-        if (blockingClass && service.classAnnotation(GrpcDotNames.NON_BLOCKING) != null) {
+    /**
+     * Generate list of {@link ClassInfo} with {@code service} as the first element and the class implementing
+     * {@code io.grpc.BindableService} (the protobuf generated {@code *ImplBase}) as the last one.
+     */
+    private static List<ClassInfo> classHierarchy(ClassInfo service, IndexView index) {
+        List<ClassInfo> collected = new ArrayList<>();
+        for (ClassInfo ci = service; ci != null;) {
+            collected.add(ci);
+
+            // Stop at the class that implements io.grpc.BindableService or
+            if (ci.interfaceNames().contains(GrpcDotNames.BINDABLE_SERVICE)) {
+                break;
+            }
+
+            DotName superName = ci.superName();
+            if (superName == null) {
+                break;
+            }
+            ci = index.getClassByName(superName);
+        }
+        return collected;
+    }
+
+    private static Set<DotName> relevantAnnotations(List<ClassInfo> classHierarchy) {
+        Set<DotName> relevant = new HashSet<>();
+        relevant.add(GrpcDotNames.BLOCKING);
+        relevant.add(GrpcDotNames.NON_BLOCKING);
+        relevant.add(TRANSACTIONAL);
+
+        return classHierarchy.stream()
+                .flatMap(ci -> ci.classAnnotations().stream())
+                .map(AnnotationInstance::name)
+                .filter(relevant::contains)
+                .collect(Collectors.toSet());
+    }
+
+    private static Set<DotName> relevantAnnotations(List<ClassInfo> classHierarchy, MethodInfo method) {
+        Set<DotName> relevant = new HashSet<>();
+        relevant.add(GrpcDotNames.BLOCKING);
+        relevant.add(GrpcDotNames.NON_BLOCKING);
+        relevant.add(TRANSACTIONAL);
+
+        return classHierarchy.stream()
+                .map(ci -> ci.method(method.name(), method.parameters().toArray(new Type[0])))
+                .filter(Objects::nonNull)
+                .flatMap(mi -> mi.annotations().stream())
+                .map(AnnotationInstance::name)
+                .filter(relevant::contains)
+                .collect(Collectors.toSet());
```

Comment:
Not sure whether I understand correctly. But it's an important aspect.

The idea here is to merge all `@Blocking`, `@NonBlocking` and `@Transactional` annotations from the whole hierarchy. But that might not be what a user would expect.

Are you proposing to only use the annotations on the "top-most" declaration - i.e. the actual service implementation (having the `@GrpcService` annotation) and for methods the annotations from the "top-most" implementation/override?
Not sure how to exactly implement it here though - if we'd have `@Blocking`/`@NonBlocking` only on methods, it can certainly use only the annotations on the "top-most" method implementation/override. It becomes tricky if a method's explicitly declared e.g. `@NonBlocking` in a base class but the actual service implementation class has a `@Blocking` annotation. I tend to use the annotation from the method, as it's more specific.

For example:
```
@GrpcService @NonBlocking
class FooService extend BaseFooService {
  // bar() would be Blocking here
}

class BaseFooService extends FooImplBase {
  @Override
  @Blocking
  void bar() {}
}
```

```
@GrpcService @NonBlocking
class FooService extend BaseFooService {
  // bar() would be NonBlocking here
}

@Blocking
class BaseFooService extends FooImplBase {
  @Override
  void bar() {}
}
```

Reply:
hmm... Maybe, for methods, to avoid confusion we should only apply blocking/nonblocking on the actual implementation, so if a method is inherited and not overriden, use the annotation from it, if it is overriden, apply only the annotation from this method (if there is one, none if there's none)?

The class level annotation could still be inheritted (the "closest" one to the implementation).

@mkouba WDYT?

Reply:
I've pushed another commit with an updated logic (probably while you've been writing the comment).
The "rule" when a method is considered blocking is described here https://github.com/quarkusio/quarkus/pull/23927/commits/ae3cf7b8074f4aa1656cbdd3e74dd15c6bdf95bb#diff-950bac0249018afb3983346929ea669c55e8d151de6161dc085a39b842ddd5c1R361

Reply:
Note: the behavior for existing implementations doesn't change - but I hope it's more intuitive when there are overridden methods.